lib/spack/spack/solver/asp.py

```diff
@@ -2026,33 +2037,80 @@ def _develop_specs_from_env(spec, env):
     spec.constrain(dev_info['spec'])
 
 
-#
-# These are handwritten parts for the Spack ASP model.
-#
-def solve(specs, dump=(), models=0, timers=False, stats=False, tests=False,
-          reuse=False):
-    """Solve for a stable model of specs.
-
-    Arguments:
-        specs (list): list of Specs to solve.
-        dump (tuple): what to dump
-        models (int): number of models to search (default: 0)
+class Solver(object):
+    """This is the main external interface class for solving.
+
+    It manages solver configuration and preferences in once place. It sets up the solve
+    and passes the setup method to the driver, as well.
+
+    Properties of interest:
+
+      ``reuse (bool)``
+        Whether to try to reuse existing installs/binaries
+
+      ``minimal (bool)``
+        If ``True`` make minimizing nodes the top priority, even higher
+        than defaults from packages and preferences.
```

Comment:
this doesn't quite match the description you gave in the PR, it sounds like it makes minimizing builds the priority -- I think we need to make sure to document which it is properly.

if reuse is on, minimize is on, and cmake+openssl is already installed, is `spack spec cmake` supposed to use the old one?